safeeyes/plugins/smartpause/plugin.py

```diff
@@ -126,93 +174,102 @@ def init(ctx, safeeyes_config, plugin_config):
-                logging.info('Resume Safe Eyes due to user activity')
-                smart_pause_activated = False
-                idle_period = (datetime.datetime.now() - idle_start_time)
-                idle_seconds = idle_period.total_seconds()
-                context['idle_period'] = idle_seconds
-                if interpret_idle_as_break and idle_seconds >= next_break_duration:
-                    # User is idle for break duration and wants to consider it as a break
-                    enable_safe_eyes()
-                elif idle_seconds < break_interval:
-                    # Credit back the idle time
-                    next_break = next_break_time + idle_period
-                    enable_safe_eyes(next_break.timestamp())
-                else:
-                    # User is idle for more than the time between two breaks
-                    enable_safe_eyes()
+
+    if not __is_active():
+        return False  # stop the timeout handler.
+
+    system_idle_time = idle_checker.idle_seconds()
+    if system_idle_time >= idle_time and context['state'] == State.WAITING:
+        smart_pause_activated = True
+        idle_start_time = datetime.datetime.now()
+        logging.info('Pause Safe Eyes due to system idle')
+        disable_safe_eyes(None)
+    elif system_idle_time < idle_time and context['state'] == State.STOPPED:
+        logging.info('Resume Safe Eyes due to user activity')
+        smart_pause_activated = False
+        idle_period = (datetime.datetime.now() - idle_start_time)
+        idle_seconds = idle_period.total_seconds()
+        context['idle_period'] = idle_seconds
+        if interpret_idle_as_break and idle_seconds >= next_break_duration:
+            # User is idle for break duration and wants to consider it as a break
+            enable_safe_eyes()
+        elif idle_seconds < break_interval:
+            # Credit back the idle time
+            next_break = next_break_time + idle_period
+            enable_safe_eyes(next_break.timestamp())
+        else:
+            # User is idle for more than the time between two breaks
+            enable_safe_eyes()
+
+    return True  # keep this timeout handler registered.
 
 
 def on_start():
     """
-    Start a thread to continuously call xprintidle.
+    Begin polling to check user idle time.
     """
-    global active
-    if not __is_active():
+    global idle_checker
+    global _timer_event_id
+
+    if __is_active():
         # If SmartPause is already started, do not start it again
-        logging.debug('Start Smart Pause plugin')
-        __set_active(True)
-        Utility.start_thread(__start_idle_monitor)
+        return
+
+    logging.debug('Start Smart Pause plugin')
+    idle_checker = idle_checker_for_platform(context)
+
+    __set_active(True)
+
+    # FIXME: need to make sure that this gets updated if the waiting_time config changes
+    _timer_event_id = GLib.timeout_add_seconds(waiting_time, __idle_monitor)
```

Comment:
From reading `init()`, it looks like `waiting_time` is *almost always* 2, except in the cases where `plugin_config.idle_time` is *less* than two seconds.

Is there still a use case for that requirement? I'm having trouble imagining what that might be for.

Reply:
Users have the option to specify the minimum idle time to pause Safe Eyes (by default 5 sec). Executing xprintidle every 5 seconds may pause after  seconds. That's why I made it less than  seconds which is hard coded to 2. However if the user reduces the minimum idle time to 1 sec, waiting_time also will be 1 sec.

Waiting time is used to add a delay to the while loop so that it doesn't eat the CPU. If your solution does not need busy waiting, you can skip this.

Reply:
It does still need a `waiting_time`. The part I am unclear on is this

> However if the user reduces the minimum idle time to 1 sec, waiting_time also will be 1 sec.

What sort of user is it that wants a minimum idle time of only one second? What kind of workflow do they have? How would they be troubled if their `waiting_time` never went below two seconds?

It must be very different than how I have considered using this, because I've set my minimum `idle_time` much higher than that.

Reply:
I do not have any specific user requirements. Personally I set it to 5 seconds but in I didn't want to deal with issues complaining people about smart pause not respecting/not allowing a minimal waiting time. That's why I try my best to keep all settings configurable and let people to choose.

However your argument makes sense and we can set the minimum bound to be 5. Just checked the `config.json` it is already set to a minimum of 5. So we can go with a minimum of 5 seconds idle time and a waiting time slightly less than that.